core/src/main/java/hudson/slaves/NodeProvisioner.java

```diff
@@ -217,29 +218,43 @@ public void run() {
-                                Node node = f.future.get();
-                                for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
-                                    cl.onComplete(f, node);
-                                }
-    
-                                jenkins.addNode(node);
-                                LOGGER.log(Level.INFO,
-                                        "{0} provisioning successfully completed. " 
-                                                + "We have now {1,number,integer} computer(s)",
-                                        new Object[]{f.displayName, jenkins.getComputers().length});
-                            } catch (InterruptedException e) {
-                                throw new AssertionError(e); // since we confirmed that the future is already done
-                            } catch (ExecutionException e) {
-                                LOGGER.log(Level.WARNING, "Provisioned agent " + f.displayName + " failed to launch",
-                                        e.getCause());
-                                for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
-                                    cl.onFailure(f, e.getCause());
+                                Node node = null;
+                                try {
+                                    node = f.future.get();
+                                } catch (InterruptedException e) {
+                                    throw new AssertionError("InterruptedException occurred", e); // since we confirmed that the future is already done
+                                } catch (ExecutionException e) {
+                                    Throwable cause = e.getCause();
+                                    if(!(cause instanceof AbortException)) {
+                                        LOGGER.log(Level.WARNING, "Unexpected exception encountered while "
+                                                + "provisioning agent " + f.displayName, cause);
+                                    }
+                                    for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+                                        cl.onFailure(f, cause);
+                                    }
+                                    node = null;
                                 }
-                            } catch (IOException e) {
-                                LOGGER.log(Level.WARNING, "Provisioned agent " + f.displayName + " failed to launch", 
-                                        e);
-                                for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
-                                    cl.onFailure(f, e);
+
+                                if (node != null) {
+                                    for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+                                        cl.fireOnComplete(f, node);
```

Comment:
I think adding new events after the save is the way to go. The original documented contract implies that the existing callbacks will all be only after the `Future` completes, so there is no compatibility issue with adding `onCommitted` (signalling that Jenkins has committed ownership of the node) and `onRollback` (signalling that Jenkins has aborted ownership of the node and the node is now unmanaged) events.